### PR TITLE
Micro: create new micro-layer, add basic Sheetlet

### PR DESCRIPTION
**Pull request by DLehenbauer**
Copy of @jack-williams's basic Sheetlet implementation with tests & benchmarks to serve as a baseline.  (Currently in a separate 'micro' package)

### CODE REVIEW

File: packages/common/tool-config/package.json [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+  "name": "@tiny-calc/tool-config",
+  "version": "0.0.2",
+  "description": "Common config for build tools",
+  "sideEffects": "false",
+  "scripts": {
+    "build": "echo No build steps necessary."
+  },
+  "author": "Microsoft",
+  "license": "MIT"
+}
```
Comment (DLehenbauer) [inline packages/common/tool-config/package.json]: (The 'tool-config' is a pseudo-package used to share tsconfig.json across projects.)

File: packages/core/micro/src/sheetlet.ts [hunk 2]
```diff
@@ -0,0 +1,871 @@
...
+            case "number":
+            case "string":
+            case "boolean":
+                return value;
+            case "function":
+                return "<function>";
+            case "object":
+                if (value instanceof Range) {
+                    return this.primitiveFromValue(origin, value.read("value", origin) as CalcValue<O>);
+                }
```
Comment (DLehenbauer) [inline packages/core/micro/src/sheetlet.ts]: FYI - evaluating a formula like 'A1' would attempt to "stringify" Range.  I added this special case as a workaround.

File: packages/core/nano/src/produce.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+            : props);
+
+    for (const value of Object.values(subject)) {
+        if (typeof value === "object" && value) {
+            produce(value);
+        }
+    }
+
+    return subject as any;
+}
```
Comment (DLehenbauer) [inline packages/core/nano/src/produce.ts]: (Unrelated experiment with 'produce'...  may not keep this.)

File: packages/core/nano/src/types.ts [hunk 4]
```diff
@@ -78,17 +82,19 @@ export interface IMatrixConsumer<T> {
...
      */
     removeMatrixConsumer(consumer: IMatrixConsumer<T>): void;
 
-    openMatrix(consumer: IMatrixConsumer<T>): {
-        readonly numRows: number;
-        readonly numCols: number;
-        read(row: number, col: number): T;
-    }
+    openMatrix(consumer: IMatrixConsumer<T>): IMatrixReader<T>;
 }
```
Comment (DLehenbauer) [inline packages/core/nano/src/types.ts]: (some unrelated interface tweaks based on experiments with IProducer/IConsumer ... should keep these, provided we keep the interfaces.)

**Comment by DLehenbauer** [comment]
Change also includes some noise from:
* Unrelated experiments with produce(..) that we may or may not keep
* Unrelated tweaks to IConsumer/IProducer interfaces, based on some experimental usage.
* Some test utilities from IConsumer/IProducer interface experiments.

File: packages/core/micro/src/sheetlet.ts (hunk 2 above)
Comment (jack-williams) [reply packages/core/micro/src/sheetlet.ts]: Good catch! I had ` coerceResult` for evaluating cells but forgot to use it here. Might be possible to unify the code by making ` coerceResul` generic.

File: packages/core/nano/test/util/loggingConsumer.ts [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+export const nullConsumer: IConsumer<unknown> & IVectorConsumer<unknown> & IMatrixConsumer<unknown> = new NullConsumer<unknown>();
+
+export class LoggingConsumer<T> implements IConsumer<T>, IVectorConsumer<T>, IMatrixConsumer<T> {
+    public readonly log: any[] = [];
+
+    public static setProducerId(producer: any, value: string) { producer.id = value; }
+    private getProducerId(producer: any) { return producer.id || "Error: Missing call to LoggingConsumer.setProducerId(..)"; }
+
+    // #region IConsumer<T>
+    public valueChanged<U extends T, K extends keyof U>(property: K, value: U[K], producer: IProducer<U>): void {
```
Comment (jack-williams) [inline packages/core/nano/test/util/loggingConsumer.ts]: On a slightly related note: I found that in practice have two generic parameters for `IConsumer` was a pain. I think it was easier to jut have `IProducer<T>` and `IConsumer<T>` and rely on subtyping to allow things to be wider if needed. Something to keep an eye on through iterations.

Comment (DLehenbauer) [reply packages/core/nano/test/util/loggingConsumer.ts]: Will do, thx.

**Pull request merged**